src/cryptonote/onetime_key.rs

```diff
@@ -138,29 +141,47 @@ pub struct SubKeyChecker<'a> {
 impl<'a> SubKeyChecker<'a> {
     /// Generate the table of sub spend keys `K(S) \in major x minor` from a view pair mapped to
     /// their Sub-address indexes.
-    pub fn new(keys: &'a ViewPair, major: Range<u32>, minor: Range<u32>) -> Self {
+    pub fn new(
+        keys: &'a ViewPair,
+        major: Range<u32>,
+        minor: Range<u32>,
+    ) -> Result<Self, AddressError> {
         let mut table = HashMap::new();
-        major.for_each(|maj| {
-            minor.clone().for_each(|min| {
+
+        for maj in major {
+            for min in minor.clone() {
                 let index = Index {
                     major: maj,
                     minor: min,
                 };
-                let spend = subaddress::get_spend_public_key(keys, index);
-                table.insert(spend, index);
-            });
-        });
-        SubKeyChecker { table, keys }
+                match subaddress::get_spend_public_key(keys, index) {
+                    Ok(spend) => {
+                        table.insert(spend, index);
+                    }
+                    Err(e) => {
+                        return Err(e);
+                    }
+                }
```

Comment:
cant we just use a `?` here

Reply:
The previous `for_each` closures did not allow the different match arms, and now we need a closure that handles both the success and failure modes.